Revise the manual (#3231)
GaloisInc/saw-script: doc/saw-user-manual/the-sawscript-language.md

by using the empty escape sequence `\&`.

Text enclosed in double curly braces (`{{ }}`) is treated as Cryptol
code (in particular, a Cryptol expression) and parsed by the Cryptol
parser.
Text enclosed in curly braces and bars (`{| |}`) is parsed as a Cryptol
type.
Cryptol types are expressions at the SAWScript level, as discussed below.
:::{code-block} sawscript
   let xs = {{ [1, 2, 3] }};
   let t = {| Integer |};
:::

The following identifiers are reserved words:

:::{list-table}
* - `and`
  - `do`
  - `hiding`
  - `import`
  - `let`
  - `submodule`
  - `typedef`
* - `as`
  - `else`
  - `if`
  - `in`
  - `rec`
  - `then`
  - ` `
:::

The following identifiers for built-in types are also currently
treated as reserved words.
This may change in the future, as handling them as reserved words is
neither necessary nor particularly desirable.

:::{list-table}
* - `AIG`
  - `CrucibleSetup`
  - `JavaSetup`
  - `MIRSpec`
  - `TopLevel`
* - `Bool`
  - `Int`
  - `LLVMSetup`
  - `ProofScript`
  - `Type`
* - `CFG`
  - `JVMMethodSpec`
  - `LLVMSpec`
  - `String`
  -
* - `CrucibleMethodSpec`
  - `JVMSpec`
  - `MIRSetup`
  - `Term`

>   -

Reviewer: I'm not quite sure I understand how the identifiers here are grouped. (There appears to be some sort of nesting of `*` and `-` bullet points.)

Also, there are some unused `-` bullet points after `String` and `Term`.

Author: It creates a table in the output. I'd post a screencap, but GH doesn't allow that. (I'll send you one by a side channel.) I've added some comments.

:::

## Types

SAWScript is strongly typed, as mentioned above.
The following basic types are available:

- `Bool` (booleans; values are spelled `true` and `false`)
- `Int` (unbounded integers)
- `String` (Unicode strings)

There are also the following derived types:

- Tuples.
  Tuple types are written as tuples of types, that is, a comma-separated
  list of types in parentheses.
  An empty pair of parentheses gives the unit type.
  (The unit type has one value, also written `()`; a value of type `()`
  contains no information.
  It is like `void` in C and Java.)
  There are no monoples (tuples of arity 1).
- Lists.
  List types are written with a type name in square brackets: `[Int]` is a
  list of integers.
  Lists must be homogeneous; there is only one element type for the whole list.
- Records.
  Records are written with curly braces and a comma-separated list of

>   field names and type signatures set off with colons: `{ name: String, length: Int }`.

Reviewer: Worth mentioning that record values use slightly different syntax than record types, using equal signs instead of colons (e.g., `{ name = "hello", length = 42 }`).

Author: I guess, this whole subsection is about types and the record value syntax appears later...

  Record types are purely structural, that is, all record types with the
  same fields of the same types are the same type.
- Functions.
  Function types are written with a right-arrow `->`.
  Functions are curried, so a function of two arguments and a function
  of one argument that returns a function of one argument are the same
  type, and function types with multiple arguments have multiple arrows.

As already mentioned there are five monad types:

- `TopLevel` (the type of generic SAW actions)
- `ProofScript` (the type of proof tactic actions)
- `LLVMSetup` (the type of actions producing LLVM speciications)
- `JVMSetup` (the type of actions producing JVM speciications)

> - `MIRSetup` (the type of actions producing MIR speciications)

Reviewer 2: ```suggestion
- `LLVMSetup` (the type of actions producing LLVM specifications)
- `JVMSetup` (the type of actions producing JVM specifications)
- `MIRSetup` (the type of actions producing MIR specifications)
```

> Monad types have kind `* -> *` (meaning they take an argument) so for

Reviewer: Have we introduced the notion of kinds yet in the manual? I think saying kind `* -> *` could potentially be head-scratching for folks not in the know.

Author: I would expect most people reading this to be in the know. But maybe not. I rearranged the wording, and I also clarified two other references to kinds earlier in the manual

example the type `ProofScript Int` is a computation in the ProofScript
monad that produces an integer when run.

### Other Built-In Types

There are numerous further built-in types used for various verification
tasks.

<!-- XXX at least some of this should be shoved over to the reference -->

Cryptol-related types:

- `Term` (the type of Cryptol expressions and SAWCore value terms)
- `Type` (the type of Cryptol values and SAWCore type terms)
- `CryptolModule` (a handle for a Cryptol module; see [XXX](XXX imports) below)

Proof-related types:

- `SatResult` (the result produced by the `sat` operation)

> - `ProofResult` (the result produced by the `proof` operation)

Reviewer 2: ```suggestion
- `ProofResult` (the result produced by the `prove` operation)
```

It is not legal to write a function that uses a tuple pattern in place
of its name.

Binding a variable name that already exists will, in general, create a
new variable for the new value; the old value is left alone and
references to it are unchanged.

There is an exception to this: global variables at the syntactic top
level can be bound with the special variant syntax
`let rebindable ...`.
These variables can then be updated by a subsequent `let rebindable`;
in effect they are mutable globals.
Updates must be with the same type, which must be monomorphic.
See [Issue 1646](https://github.com/GaloisInc/saw-script/issues/1646)
for further discussion and the history of this (somewhat dubious)
feature.
Note that `rec rebindable` is not permitted, and variables bound with
`<-` cannot be rebindable either.
Rebinding a `rebindable` variable with a _non_-rebindable definition
masks it with a new immutable version, but does _not_ update it.

One can also write `let {{ ... }};`, in which the double-braces can
contain not just an expression but any Cryptol declaration or
let-binding.
This is a convenient way to insert Cryptol type declarations; it also

> allows using Cryptol elim forms that cannot readly be expressed directly in

Reviewer: The phrase "elim forms" might be a bit too jargon-y. Perhaps just replace with "definitions"?

Author: That loses the point, though, which is that you can do things like `let {{ (a, b) = ... }}`

I changed to "binding patterns"; speak up if you have a better idea.

of monad type.
The syntax is a variable name, a left arrow `<-`, and an expression.
The expression is first evaluated purely; then, roughly speaking, the
resulting monadic action is executed in its monad, producing a value
which is stored in the variable name.

Precisely speaking, the monad action is bound into the chain of
actions associated with the current do-block (the syntactic top level
being functionally an arbitrary long do-block) and the actions do not
actually _happen_ until that block, and the block it is bound into,
and so forth, and eventually the actions at the syntactic top level,
are all actually executed by the SAWScript interpreter.

If you are not familiar with the use of monads to sequence actions and
carry state around, we recommend you seek out one of the many
explanations available online.

Note however that SAWScript is unlike Haskell in that it executes
eagerly: expressions are evaluated when they are encountered, as in
conventional languages, rather than only when their values are used
somewhere.
Only the execution of monadic actions is delayed.

Monad bind can also use `_` or tuple patterns on the left hand side.

> Also, leaving the variable name and the arrow off is entirely is

Reviewer 2: ```suggestion
Also, leaving the variable name and the arrow off is entirely
```

structured way to load SAWScript code.

## Expressions

Base expressions include constants (integer constants, string constants,
the unit value `()`, and the empty list `[]`), Cryptol expressions in double
curly braces, Cryptol types enclosed in `{| |}`, variable lookups including
names of builtins, and subexpressions in parentheses.

Comma-separated lists of two or more expressions in parentheses are
tuple literals and produce tuple values.
List literals are comma-separated expressions enclosed in square
brackets `[]`.
Record literals are comma-separated lists in single braces `{ }`, where
each field is assigned with a field name, an equal sign `=`, and a
subexpression.
:::{code-block} sawscript
let data = {
   name = "John Smith",
   address = "55 Elm St.",
   city = "Portland",
   state = "Oregon",
   phone = "555-1234"
};
:::

An expression followed by a dot and an identifier looks up the
so-named field in a record value: `data.phone`.
An expression followed by a dot and an integer constant looks up the
nth field of a tuple value.
Tuple indexes are zero-based.

Juxtaposition of expressions does function application, like in Haskell

> and ML: `f x` applies `x` to `f`.

Reviewer 2: ```suggestion
and ML: `f x` applies the function `f` to an argument `x`.
```

    in particular to catch the failure of a pure function, it must be
    enclosed in a do-block when passed to `fails`.
    Otherwise the failure happens evaluating the function before `fails`
    starts.
    See [issue #2424](https://github.com/GaloisInc/saw-script/issues/2424)
    for further discussion.

### Library-Level Builtins

There is no SAWScript prelude as such, just a collection of builtins that
resemble a standard library.
Here are some of the functions available:

#### String Functions

- `show : {a} a -> String` computes the textual representation of its
argument in the same way as `print`, but instead of displaying the value
it returns it as a `String` value for later use in the program. This can
be useful for constructing more detailed messages later.

- `str_concat : String -> String -> String` concatenates two `String`
values, and can also be useful with `show`.

#### List Functions

- `concat : {a} [a] -> [a] -> [a]` takes two lists and returns the
  concatenation of the two.
  Note that this is more usually called `append`; the function
  more usually called `concat` is called `XXX` in SAWScript.

- `head : {a} [a] -> a` returns the first element of a list.

> - `tail : {a} [a] -> [a]` returns everything except the first element.

Reviewer: It might also be worth clarifying that calling `head` or `tail` on an empty list will raise an error.

Author: Fixed

- `length : {a} [a] -> Int` counts the number of elements in a list.

- `null : {a} [a] -> Bool` indicates whether a list is empty (has zero
elements).

- `nth : {a} [a] -> Int -> a` returns the element at the given position,
with `nth l 0` being equivalent to `head l`.

#### Timing Functions

- `time : {a} TopLevel a -> TopLevel a` runs any other `TopLevel`
command and prints out the time it took to execute.

- `with_time : {a} TopLevel a -> TopLevel (Int, a)` returns both the
original result of the timed command and the time taken to execute it
(in milliseconds), without printing anything in the process.

- `timeout : {a} Int -> TopLevel a -> TopLevel a` takes a timeout
  in milliseconds and an action to run.
  If the action takes longer, it is stopped and `timeout` fails.

- `timeout_handle : {a} Int -> TopLevel a -> TopLevel a -> TopLevel a` is
  like `timeout` but takes an additional action that is run if the
  operation times out.

The `timeout` and `timeout_handle` builtins, like the `fails` builtin,
can only handle timeouts in monadic actions.
If you need a timeout for a pure function, wrap it in a do-block.

#### System Functions

> - `get_opt` _`n`_ returns the current script's command-line argument
>   _`n`_.

Reviewer 2: For consistency with the other functions in this list:
```suggestion
- `get_opt : Int -> String` returns the current script's command-line argument.
```

Author: It really needs to refer to _`n`_. I think I'd rather change the others...

Author: (so I did that)
